### DISCUSSION TITLE
Validate Symfony Forms without WebTestCase

### DISCUSSION THREAD

**Discussion by mpiot** [symfony/symfony#51181] [Q&A]
Hi everyone,
How do you test your application forms ? In our case we do that via WebTestCase, but it take a lot of times (4000+ tests / 12 minutes), lot of tests are:
 - form with empty data
 - empty collections
 - test that we can't set a value greater than an other (eg: a Min and Max inputs)
 - ...

I think, there are better way to do than do "real" request to the application and check the error in returned response via the Crawler.

I've think about:
 - Unit test forms (https://symfony.com/doc/current/form/unit_testing.html), but it appears better to test Custom types
 - Just create an Entity, then validate it and check errors (maybe cumulate it with a test that check the Form validation groups)

But maybe some of you have an idea ?
Thanks a lot :slightly_smiling_face:

**Comment by MatTheCat** [comment]
Hello,

sounds like you want to test your forms *validation* so the minimum you need is [integration tests](https://symfony.com/doc/current/testing.html#integration-tests), e.g.

```php
<?php

namespace Whatever;

use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Symfony\Component\Form\FormFactoryInterface;

class YourFormTest extends KernelTestCase
{
    public function testSomeFieldIsRequired(): void
    {
        /** @var FormFactoryInterface $formFactory */
        $formFactory = static::getContainer()->get(FormFactoryInterface::class);
        $form = $formFactory->create(YourFormType::class, $yourFormData, ['csrf_protection' => false]);

        $form->submit(['someField' => null]);

        $errors = $form->getErrors(true);
        self::assertCount(1, $errors);

        // more assertions on errors
    }
}
```

Be aware this won’t test the front-end like a `WebTestCase` would do though.

Regarding tests performance, you can read https://maks-rafalko.github.io/blog/2021-11-21/symfony-tests-performance/ for some tips.

**Reply by mpiot** [reply]
Thanks for your answer :)
I'll look the link :)
